### ISSUE TITLE
What is the status of tensorflow on spack? 

### ISSUE THREAD

**Issue by llpesce** [spack/spack#9352]
I need tensorflow under my spack stack on theta@ANL

I have considered installing tensorflow outside of spack, but I would rather not.

I would not mind using an external spack package, but I am not sure which one to use so I am asking the community about what is the best candidate

**Comment by Sinan81** [comment]
Wondering if there is any update on the status of tensorflow on Spack since this issued was posted last year.

In the mean time, I am compiling it outside of Spack (with dependencies provided by spack). Not able to compile tensorflow v1.13 or 1.14 as Bazel versions beyond 17.2 is not available in Spack yet. #12398

**Comment by adamjstewart** [comment]
The closest package in Spack right now is #2043. It tries to play nicely with Bazel, but has many limitations. For example, it downloads its own set of dependencies, so it won't build on an air-gapped network. This is similar to how the Easybuild TensorFlow package works.

#3244 was a failed attempt to build TensorFlow using CMake. This was abandoned as Google decided not to support CMake, and community support would never be able to keep up with all the changes to the Bazel build.

If you want to pick up where #2043 left off, I would love to review another PR for a TF package! "When will Spack get TF" was the most common question we received at PEARC19, so we know there is a lot of demand for it. Once #3206 is merged, @tgamblin wants to offer pre-built binaries of TF for each CPU architecture, something that no other package manager in existence has!

**Comment by Sinan81** [comment]
@adamjstewart Thanks a lot for the update. I would certainly be interested in looking into #2043, however I am super busy currently. It would be a "game changer" if arch specific pre-built binaries of TF were provided by Spack.

**Comment by citibeth** [comment]
@Sinan81 Could you provide more info on your out-of-Spack build of tensorflow?  Specifically, can you list the Spack packages of tensorflow dependencies?  We should at least be able to make a tensorflow-deps meta-package.

**Comment by adamjstewart** [comment]
@citibeth The major hurdle with TF is that it tries to download its own deps, and there appears to be no way to prevent that. That's the main reason it hasn't been packages yet, although EasyBuild has it. PyTorch also downloads some of its own deps, but we have a package for that. I think we should move forward with a TF package, even if it downloads some of its own deps.

**Comment by citibeth** [comment]
I agree.

**Comment by Sinan81** [comment]
@citibeth : below are the dependencies that I use for building & running tensorflow. Actually, yesterday I created a feature branch for tensorflow packagefile enabling me to install tensorflow wheel compiled by hand: https://github.com/Sinan81/spack/tree/tensorflow 
```
    depends_on('py-pip', type='build')
    depends_on('bazel', type='build')

    # setuptools is also a run dependency!
    depends_on('py-setuptools', type=('build', 'run'))
    depends_on('python', type=('build', 'run'))
    depends_on('py-keras-applications@1.0.6', type=('build', 'run'))
    depends_on('py-keras-preprocessing@1.0.5', type=('build', 'run'))
    depends_on('py-six', type=('build', 'run'))
    depends_on('cuda', type=('build', 'run'))
    depends_on('cudnn', type=('build', 'run'))
    depends_on('py-numpy', type=('build', 'run'))
    depends_on('py-future@0.17.1:', type=('build', 'run'))

    # run dependencies
    depends_on('py-protobuf', type='run')
    depends_on('py-absl-py', type='run')
    depends_on('py-enum34', type='run', when='^python@2')
    depends_on('py-pbr', type='run')
    depends_on('py-funcsigs', type='run')
    depends_on('py-mock', type='run')
```

**Comment by Sinan81** [comment]
I should also add that latest version of bazel available via spack conflicts with gcc versions beyond 7.4.0.
Speaking of bazel, with the latest version available via spack, 0.17.2, we are only able to compile tensorflow version 1.12.3. It would be a huge help if newer versions of Bazel are provided by spack.

**Comment by Sinan81** [comment]
@citibeth Tested build configurations listed at the bottom of the the following page should be useful for choosing a suitable cuda, cudnn, bazel version combination: https://www.tensorflow.org/install/source

**Comment by healther** [comment]
For what it's worth: [This is our package](https://github.com/electronicvisions/spack/blob/visionary/var/spack/repos/builtin/packages/tensorflow/package.py) that we are using internally. It works but it really isn't pretty and requires the `spack install tensorflow` command to be run from somewhere with internet access. 

We always wanted to put it upstream, but things keep coming up that push that down on our priority list...

**Comment by adamjstewart** [comment]
@healther that doesn't look that bad. It's better than no package. Does anyone have the time to steal that recipe and make a PR?

**Comment by healther** [comment]
There is a graduate school next week at our institute, I might find some time at the lecture.. no promises though :)

**Comment by Sinan81** [comment]
@healther I am having problems in building bazel@0.19.0 with spack on a Centos7 machine. I am wondering if you might have some tips for it.

**Comment by healther** [comment]
@Sinan81 unfortunately I'm not really sure what exactly we had to do, if you open an issue and post some more context to your problem, I may be able to dig out some stuff...

**Comment by adamjstewart** [comment]
@healther @Sinan81 I believe someone just needs to port the existing patches to the latest version of Bazel, should be trivial.

**Comment by jxy** [comment]
As for building tensorflow from an air gapped system, you may get some ideas from how the freebsd port builds it, https://github.com/freebsd/freebsd-ports/tree/master/science/py-tensorflow

The extra distfiles that bazel needs can be separately fetched, https://github.com/freebsd/freebsd-ports/blob/master/science/py-tensorflow/Makefile.MASTER_SITES

**Comment by mslacken** [comment]
As we have tensorflow in openSUSE I just wanted to point you to the spec file we are using for that:
https://build.opensuse.org/package/show/science:machinelearning/tensorflow
There might be some tricks inside it, you can use in spack, too.

**Comment by tgamblin** [comment]
Hey folks -- if you're still interested in this note that @adamjstewart and @dodo47 merged a TensorFlow package in #13112 -- thanks very much to them for making this happen!
